library/alloc/src/vec.rs

```diff
@@ -2188,28 +2188,32 @@ impl<T> Drop for InPlaceDrop<T> {
 
 impl<T> SpecFromIter<T, IntoIter<T>> for Vec<T> {
     fn from_iter(iterator: IntoIter<T>) -> Self {
-        // A common case is passing a vector into a function which immediately
-        // re-collects into a vector. We can short circuit this if the IntoIter
-        // has not been advanced at all.
-        // When it has been advanced We can also reuse the memory and move the data to the front.
-        // But we only do so when the resulting Vec wouldn't have more unused capacity
-        // than creating it through the generic FromIterator implementation would. That limitation
-        // is not strictly necessary as Vec's allocation behavior is intentionally unspecified.
-        // But it is a conservative choice.
+        // It is common to pass a vector to a function which immediately
+        // re-collects as a vector. We can short circuit this behavior if
+        // the IntoIter has not been advanced.
+        //
+        // We reuse the original buffer if the iterator has not been advanced
+        // or if there is a small unused capacity if we reuse the buffer.
         let has_advanced = iterator.buf.as_ptr() as *const _ != iterator.ptr;
         if !has_advanced || iterator.len() >= iterator.cap / 2 {
             unsafe {
                 let it = ManuallyDrop::new(iterator);
+                // If it has been advanced, we reuse the memory by moving
+                // the data to the front.
                 if has_advanced {
                     ptr::copy(it.ptr, it.buf.as_ptr(), it.len());
                 }
                 return Vec::from_raw_parts(it.buf.as_ptr(), it.len(), it.cap);
             }
         }
 
+        // We use a conservative method to grow a new Vec with the generic
+        // from_iter if there is a large unused capacity after the
+        // optimization above. This limitation is not strictly necessary as
```

Comment:
This comment feels *more* confusing to me to be placed here, and saying "if there is a large unused capacity after the optimization above" to me reads like we're somehow realloc'ing the buffer we copied data around in which is not true here.

Reply:
"grow a new Vec with the generic from_iter" is the action and "there is a large unused capacity after the optimization above" is the condition. It does seemed clear to me, any idea to improve it?